File: packages/ubaa_app/lib/src/bridge/read/bykc.dart
part of '../bridge_backend.dart';

Future<FeatureResult> _loadBykcFeature(
  BridgeBackend backend,
  FeatureId feature,
  FeatureQuery query,
) async {
  final client = backend.client;
  switch (feature) {
    case FeatureId.bykc:
      switch (query.view) {
        case FeatureQueryView.summary:
          final result = await client.bykcCourses(
            // Core 的分页合同为 1-based；UI 的通用查询默认从 0 开始，
            // 因此这里在 bridge 边界显式收敛为首个有效页。
            page: query.page <= 0 ? 1 : query.page,
            size: query.size.clamp(1, 100),
            all: true,
          );
          final details = result.data.content
              .map(
                (item) => FeatureDetail(
                  title: item.courseName,
                  subtitle: item.courseTeacher,
                  actions: <FeatureAction>[
                    BykcSelectAction(
                      courseId: item.id,
                      eligibility: _toBykcActionEligibility(
                        item.selectEligibility,
                      ),
                    ),
                    BykcDeselectAction(
                      courseId: item.id,
                      eligibility: _toBykcActionEligibility(
                        item.deselectEligibility,
                      ),
                    ),
                  ],
                  fields: _compactFields(<FeatureField?>[
                    _field('课程 ID', item.id.toString()),
                    _field('地点', item.coursePosition),
                    _field('课程分类', item.subCategoryName),
                    _field('适用校区', _joinBykcValues(item.audienceCampuses)),
                    _field('状态', item.status.name),
                    item.courseCurrentCount == null
                        ? null
                        : _field('已选人数', '${item.courseCurrentCount}'),
                    item.courseMaxCount == null
                        ? null
                        : _field('容量', '${item.courseMaxCount}'),
                    _field('选课截止', item.courseSelectEndDate),
                    _field('退选截止', item.courseCancelEndDate),
                  ]),
                ),
              )
              .toList(growable: false);
          return _countResult(
            result.data.content.length,
            '门博雅课程',
            details: details,
            pagination: _pagination(
              page: result.data.number,
              size: result.data.size,
              total: result.data.totalElements,
              totalPages: result.data.totalPages,
            ),
            resolvedRoute: _toConnectionMode(result.route.resolvedRoute),
          );
        case FeatureQueryView.bykcDetail:
          final id = _requiredPositiveQueryInt(query.courseId, '课程 ID');
          final result = await client.bykcCourseDetail(id: id);
          final item = result.data;
          return FeatureResult.success(
            summary: '博雅课程详情',
            details: <FeatureDetail>[
              FeatureDetail(
                title: item.courseName,
                subtitle: item.courseTeacher,
                actions: <FeatureAction>[
                  BykcSelectAction(
                    courseId: item.id,
                    eligibility: _toBykcActionEligibility(
                      item.selectEligibility,
                    ),
                  ),
                  BykcDeselectAction(
                    courseId: item.id,
                    eligibility: _toBykcActionEligibility(
                      item.deselectEligibility,
                    ),
                  ),
                ],
                fields: _compactFields(<FeatureField?>[
                  _field('课程 ID', item.id.toString()),
                  _field('地点', item.coursePosition),
                  _field('开课单位', item.organizerCollegeName),
                  _field(
                    '课程分类',
                    _joinBykcValues(<String?>[
                      item.categoryName,
                      item.subCategoryName,
                    ]),
                  ),
                  _field('状态', item.status.name),
                  item.courseCurrentCount == null
                      ? null
                      : _field('已选人数', '${item.courseCurrentCount}'),
                  item.courseMaxCount == null
                      ? null
                      : _field('容量', '${item.courseMaxCount}'),
                  _field('开始', item.courseStartDate),
                  _field('结束', item.courseEndDate),
                  _field('选课开始', item.courseSelectStartDate),
                  _field('选课截止', item.courseSelectEndDate),
                  _field('退选截止', item.courseCancelEndDate),
                  _field('适用校区', _joinBykcValues(item.audienceCampuses)),
                  _field('适用学院', _joinBykcValues(item.audienceColleges)),
                  _field('适用年级', _joinBykcValues(item.audienceTerms)),
                  _field('适用人群', _joinBykcValues(item.audienceGroups)),
                  _field('联系人', item.courseContact),
                  _field('联系电话', item.courseContactMobile),
                  _field('课程简介', item.courseDesc),
                  _field('签到类型', item.courseSignType?.toString()),
                  _field(
                    '已选',
                    item.selected == null
                        ? null
                        : item.selected!
                        ? '是'
                        : '否',
                  ),
                ]),
              ),
            ],
            resolvedRoute: _toConnectionMode(result.route.resolvedRoute),
          );
        case FeatureQueryView.bykcProfile:
          final result = await client.bykcProfile();
          final item = result.data;
          return FeatureResult.success(
            summary: '博雅个人资料',
            details: <FeatureDetail>[
              FeatureDetail(
                title: item.realName ?? '博雅个人资料',
                fields: _compactFields(<FeatureField?>[
                  _field('用户 ID', item.id.toString()),
                  _field('姓名', item.realName),
                  _field('学号', item.studentNo),
                  _field('学院', item.collegeName),
                ]),
              ),
            ],
            resolvedRoute: _toConnectionMode(result.route.resolvedRoute),
          );
        case FeatureQueryView.bykcChosenCourses:
          final result = await client.bykcChosenCourses();
          final details = result.data
              .map((item) {
                final signEligibility = _toBykcActionEligibility(
                  item.signEligibility,
                );
                final signOutEligibility = _toBykcActionEligibility(
                  item.signOutEligibility,
                );
                final requiresCoordinates = _requiresBykcCoordinates(
                  item.signConfig,
                );
                return FeatureDetail(
                  title: item.courseName,
                  subtitle: item.courseTeacher,
                  actions: <FeatureAction>[
                    BykcDeselectAction(
                      courseId: item.courseId,
                      eligibility: _toBykcActionEligibility(
                        item.deselectEligibility,
                      ),
                    ),
                    BykcSignAction(
                      courseId: item.courseId,
                      kind: BykcSignKind.signIn,
                      eligibility: signEligibility,
                      requiresCoordinates: requiresCoordinates,
                    ),
                    BykcSignAction(
                      courseId: item.courseId,
                      kind: BykcSignKind.signOut,
                      eligibility: signOutEligibility,
                      requiresCoordinates: requiresCoordinates,
                    ),
                  ],
                  fields: _compactFields(<FeatureField?>[
                    _field('课程 ID', item.courseId.toString()),
                    _field('地点', item.coursePosition),
                    _field('开始', item.courseStartDate),
                    _field('结束', item.courseEndDate),
                    _field('签到状态', item.checkin?.toString()),
                    _field('成绩', item.score?.toString()),
                    _field(
                      '通过',
                      item.pass == null
                          ? null
                          : item.pass! > 0
                          ? '是'
                          : '否',
                    ),
                    _field('可签到', _bykcEligibilityLabel(signEligibility)),
                    _field('可签退', _bykcEligibilityLabel(signOutEligibility)),
                    _field(
                      '签到时间',
                      _timeWindow(
                        item.signConfig?.signStartDate,
                        item.signConfig?.signEndDate,
                      ),
                    ),
                    _field(
                      '签退时间',
                      _timeWindow(
                        item.signConfig?.signOutStartDate,
                        item.signConfig?.signOutEndDate,
                      ),
                    ),
                    _field('位置要求', _locationRequirement(item.signConfig)),
                    _field('签到类型', item.courseSignType?.toString()),
                  ]),
                );
              })
              .toList(growable: false);
          return _countResult(
            details.length,
            '门已选博雅课程',
            details: details,
            resolvedRoute: _toConnectionMode(result.route.resolvedRoute),
          );
        case FeatureQueryView.bykcStatistics:
          final result = await client.bykcStatistics();
          final details = result.data.categories
              .map(
                (item) => FeatureDetail(
                  title: item.categoryName ?? item.subCategoryName ?? '分类',
                  subtitle: item.subCategoryName,
                  fields: _compactFields(<FeatureField?>[
                    _field('要求数量', item.requiredCount?.toString()),
                    _field('通过数量', item.passedCount?.toString()),
                    _field(
                      '达标',
                      item.qualified == null
                          ? null
                          : item.qualified!
                          ? '是'
                          : '否',
                    ),
                  ]),
                ),
              )
              .toList(growable: false);
          return _countResult(
            details.length,
            result.data.totalValidCount == null
                ? '博雅修读统计'
                : '有效课程 ${result.data.totalValidCount}',
            details: details,
            resolvedRoute: _toConnectionMode(result.route.resolvedRoute),
          );
        default:
          throw const BackendException(UbaaErrorCode.invalidInput);
      }
    default:
      throw StateError('unexpected feature: $feature');
  }
}

String? _joinBykcValues(Iterable<String?> values) {
  final normalized = values
      .whereType<String>()
      .map((value) => value.trim())
      .where((value) => value.isNotEmpty)
      .toList(growable: false);
  return normalized.isEmpty ? null : normalized.join(' / ');
}

ActionEligibility _toBykcActionEligibility(
  BridgeActionEligibility eligibility,
) => switch (eligibility) {
  BridgeActionEligibility.allowed => ActionEligibility.allowed,
  BridgeActionEligibility.denied => ActionEligibility.denied,
  BridgeActionEligibility.unknown => ActionEligibility.unknown,
};

String _bykcEligibilityLabel(ActionEligibility eligibility) =>
    switch (eligibility) {
      ActionEligibility.allowed => '是',
      ActionEligibility.denied => '否',
      ActionEligibility.unknown => '未知',
    };

String? _timeWindow(String? start, String? end) {
  final normalizedStart = start?.trim();
  final normalizedEnd = end?.trim();
  if (normalizedStart == null || normalizedStart.isEmpty) return null;
  if (normalizedEnd == null || normalizedEnd.isEmpty) return normalizedStart;
  return '$normalizedStart–$normalizedEnd';
}

String? _locationRequirement(BridgeBykcSignConfig? config) {
  if (config == null) return '位置配置未知（需获取当前位置）';
  if (config.signPoints.isEmpty) return '需获取当前位置（未返回可用签到范围）';
  if (_requiresBykcCoordinates(config)) return '需获取当前位置（签到范围不完整）';
  return '指定位置（${config.signPoints.length} 处）';
}

bool _requiresBykcCoordinates(BridgeBykcSignConfig? config) {
  if (config == null || config.signPoints.isEmpty) return true;
  return config.signPoints.any(
    (point) =>
        !point.lat.isFinite ||
        !point.lng.isFinite ||
        !point.radius.isFinite ||
        point.lat < -90 ||
        point.lat > 90 ||
        point.lng < -180 ||
        point.lng > 180 ||
        point.radius <= 0,
  );
}

File: crates/ubaa-core/src/domain/bykc.rs
use serde::{Deserialize, Serialize};

use super::ActionEligibility;

/// 博雅用户资料。
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BykcUserProfile {
    pub id: i64,
    pub employee_id: Option<String>,
    pub real_name: Option<String>,
    pub student_no: Option<String>,
    pub college_name: Option<String>,
}

/// 博雅课程列表项。
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BykcCourse {
    pub id: i64,
    pub course_name: String,
    pub course_position: Option<String>,
    pub course_teacher: Option<String>,
    pub organizer_college_name: Option<String>,
    pub category_name: Option<String>,
    pub sub_category_name: Option<String>,
    pub course_contact: Option<String>,
    pub course_contact_mobile: Option<String>,
    pub course_desc: Option<String>,
    pub audience_campuses: Vec<String>,
    pub audience_colleges: Vec<String>,
    pub audience_terms: Vec<String>,
    pub audience_groups: Vec<String>,
    pub course_start_date: Option<String>,
    pub course_end_date: Option<String>,
    pub course_select_start_date: Option<String>,
    pub course_select_end_date: Option<String>,
    pub course_cancel_end_date: Option<String>,
    pub course_max_count: Option<i32>,
    pub course_current_count: Option<i32>,
    pub course_sign_type: Option<i32>,
    pub status: BykcCourseStatus,
    pub selected: Option<bool>,
    /// 当前课程的选课资格；`Unknown` 必须按拒绝处理。
    pub select_eligibility: ActionEligibility,
    /// 当前课程的退选资格；`Unknown` 必须按拒绝处理。
    pub deselect_eligibility: ActionEligibility,
}

/// 博雅课程状态。
#[derive(Clone, Copy, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum BykcCourseStatus {
    /// 课程已经开始。
    Expired,
    /// 当前用户已经选择课程。
    Selected,
    /// 尚未开始选课。
    Preview,
    /// 选课已经结束。
    Ended,
    /// 课程人数已满。
    Full,
    /// 当前可以选课。
    #[default]
    Available,
}

/// 博雅课程分页结果。
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BykcCoursePage {
    pub content: Vec<BykcCourse>,
    pub total_elements: i32,
    pub total_pages: i32,
    pub size: i32,
    pub number: i32,
}

/// 博雅已选课程记录。
#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BykcChosenCourse {
    pub id: i64,
    pub course_id: i64,
    pub course_name: String,
    pub course_position: Option<String>,
    pub course_teacher: Option<String>,
    pub course_start_date: Option<String>,
    pub course_end_date: Option<String>,
    pub select_date: Option<String>,
    pub course_cancel_end_date: Option<String>,
    pub category: Option<BykcCourseCategory>,
    pub sub_category: Option<BykcCourseSubCategory>,
    /// 原始考勤状态；缺失时不得推断为未签到。
    pub checkin: Option<i32>,
    pub score: Option<i32>,
    pub pass: Option<i32>,
    /// 当前课程的签到资格；`Unknown` 必须按拒绝处理。
    pub sign_eligibility: ActionEligibility,
    /// 当前课程的签退资格；`Unknown` 必须按拒绝处理。
    pub sign_out_eligibility: ActionEligibility,
    /// 当前已选课程的退选资格；目标始终是内层课程标识。
    pub deselect_eligibility: ActionEligibility,
    pub sign_config: Option<BykcSignConfig>,
    pub course_sign_type: Option<i32>,
    pub homework: Option<String>,
    pub homework_attachment_name: Option<String>,
    pub homework_attachment_path: Option<String>,
    pub sign_info: Option<String>,
}

/// 博雅课程一级分类。
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub enum BykcCourseCategory {
    /// 博雅课程。
    #[serde(rename = "博雅课程")]
    Boya,
    /// 未识别的一级分类。
    #[serde(rename = "未知分类")]
    Unknown,
}

/// 博雅课程二级分类。
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub enum BykcCourseSubCategory {
    /// 德育。
    #[serde(rename = "德育")]
    Moral,
    /// 美育。
    #[serde(rename = "美育")]
    Aesthetic,
    /// 劳动教育。
    #[serde(rename = "劳动教育")]
    Labor,
    /// 安全健康。
    #[serde(rename = "安全健康")]
    SafetyHealth,
    /// 其他方面。
    #[serde(rename = "其他方面")]
    Other,
    /// 未识别的二级分类。
    #[serde(rename = "未知类型")]
    Unknown,
}

/// 博雅签到配置。
#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BykcSignConfig {
    pub sign_start_date: Option<String>,
    pub sign_end_date: Option<String>,
    pub sign_out_start_date: Option<String>,
    pub sign_out_end_date: Option<String>,
    pub sign_points: Vec<BykcSignPoint>,
}

/// 博雅签到地理位置。
#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
pub struct BykcSignPoint {
    pub lat: f64,
    pub lng: f64,
    pub radius: f64,
}

/// 博雅统计明细。
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BykcStatistic {
    pub category_name: Option<String>,
    pub sub_category_name: Option<String>,
    pub required_count: Option<i32>,
    pub passed_count: Option<i32>,
    pub qualified: Option<bool>,
}

/// 博雅统计结果。
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BykcStatistics {
    pub total_valid_count: Option<i32>,
    pub categories: Vec<BykcStatistic>,
}

/// 博雅课程写操作结果。
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BykcActionResult {
    pub message: String,
}

/// 博雅签到/签退请求。
#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BykcSignRequest {
    pub course_id: i64,
    pub lat: Option<f64>,
    pub lng: Option<f64>,
    pub sign_type: i32,
}

/// 博雅签到预检确认摘要。
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BykcSignPreflight {
    pub course_id: i64,
    pub course_name: String,
    pub sign_type: i32,
    pub window_start: String,
    pub window_end: String,
    pub location_requirement: BykcSignLocationRequirement,
}

/// 博雅签到本次提交使用的位置来源。
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum BykcSignLocationRequirement {
    /// Core 将从上游给出的签到范围生成本次坐标。
    ConfiguredRange,
    /// 上游完整签到点列表不能保证每次都可生成坐标，调用方已提供安全回退坐标。
    ProvidedCoordinates,
}
